rfcs/0127-issues-warnings.md

```diff
@@ -0,0 +1,199 @@
+---
+feature: issues-warnings
+start-date: 2022-06-11
+author: piegames
+co-authors: —
+shepherd-team: @lheckemann, @mweinelt, @fgaz
+shepherd-leader: @mweinelt
+related-issues: https://github.com/NixOS/nixpkgs/pull/177272
+---
+
+# RFC: Nixpkgs issues and warnings
+
+## Summary
+[summary]: #summary
+
+Inspired by the derivation checks for broken and insecure packages, a new system called "warnings" is introduced. It is planned to eventually replace the previously mentioned systems, as well as the current – undocumented – "warnings" (which currently only prints a trace message for unmaintained packages). `meta.issues` is added to derivations, which can be used to let the evaluation of individual packages fail with a custom message. This will then be used to warn users about packages that are in need of maintenance. Packages that have an open issue for a long time should eventually be removed, although doing so is not part of the RFC.
+
+## Motivation
+[motivation]: #motivation
+
+Nixpkgs has the problem that it is often treated as "append-only", i.e. packages only get added but not removed. There are a lot of packages that are broken for a long time, have end-of-life dependencies with known security vulnerabilities or that are otherwise unmaintained.
+
+Let's take the end of life of Python 2 as an example. (This applies to other ecosystems as well, and will come up again and again in the future.) It has sparked a few bulk package removal actions by dedicated persons, but those are pretty work intensive and prone to burn out maintainers. A goal of this RFC is to provide a way to notify all users of a package about the outstanding issues. This will hopefully draw more attention to abandoned packages, and spread the work load. It can also help soften the removal of packages by providing a period for users to migrate away at their own pace.
+
+Apart from that, there is need for a general per-package warning mechanism in nixpkgs – one that is stronger than `builtins.trace`.
+
+## Detailed design
+[design]: #detailed-design
+
+### Package issues
+
+A new attribute is added to the `meta` section of a package: `issues`. If present, it is a list of attrsets which each have at least the following fields:
+
+- `kind`: Required. If present, the resulting warning will be printed as `kind: message`.
+- `message`: Required. A string message describing the issue with the package.
+- `name`: Optional but recommended. Give the issue a custom name for more easy filtering
+- `date`: Required. An ISO 8601 `yyyy-mm-dd`-formatted date from when the issue was added.
+- `urls`: Optional, list of strings. Can be used to link issues, pull requests and other related items.
+- `resolved`: Optional. Indicate that the issue is resolved so that it does not generate a warning anymore.
```

Comment:
This sounds like a bad idea and I think it's the wrong solution to the original problem it intends to solve, which I think is "ignoring all issues for a package that doesn't have any issues anymore, but then gets more issues in the future". A better solution is to not allow wildcards at least for the issue kind and require people to explicitly ignore individual ones.

But even better: Don't allow filtering by issue kind, but require filtering by issue name, which works well with changing `meta.issues` to be an attribute set with a required name.

Reply:
> This sounds like a bad idea and I think it's the wrong solution to the original problem it intends to solve, which I think is "ignoring all issues for a package that doesn't have any issues anymore, but then gets more issues in the future".

No this is not the intended use case. The intended use case is "hint users to clean up their list of ignored issues once they are resolved and thus not needed anymore".

But now that you point this out, I notice that this will not work the way I intended if wildcards are used. Maybe this attribute should better be removed again then.

Reply:
I removed the "resolved" attribute